library/std/src/net/ip.rs

```diff
@@ -79,10 +78,10 @@ pub enum IpAddr {
 /// assert_eq!("127.0.0.1".parse(), Ok(localhost));
 /// assert_eq!(localhost.is_loopback(), true);
 /// ```
-#[derive(Copy)]
+#[derive(Copy, Clone, PartialEq, Eq, Hash)]
 #[stable(feature = "rust1", since = "1.0.0")]
 pub struct Ipv4Addr {
-    inner: c::in_addr,
+    octets: [u8; 4],
```

Comment:
Why not use a u32 here?

Reply:
I always wondered the same question. Funamentally an IPv4 address is a 32-bit integer, and an IPv6 is a 128-bit integer. For IPv4 especially, these integers don't have hardcoded semantics regarding their position, as ever since the custom netmask days you could arbitrarily split it into a network/host component, e.g. 1.2.3.4/27. So splitting into octets of 8 bits is arbitrary and only useful for presentation, which should be done with Display or some other trait, rather than the core representation.

IPv6 may be more complex in case different parts have more fixed semantics, but even then it's different than 16 octets of 8 bits.

Reply:
I'd welcome benchmarks showing which representation is more efficient performance wise. Most of the methods on `Ipv4Addr` work on octets though. Even the constructor happens to take the four octets separately, and not a single 32 bit integer. As a result most of the implementation was shorter and cleaner when represented as `[u8; 4]`.

If we were to represent it as a `u32` I have no doubt there would be a long debate about whether or not it should be stored as host byte order or network byte order. If going for a "ideal Rust layout" I'd definitely argue for host byte order though.

I don't have a super strong opinion on `[u8; 4]` vs `u32`. But this at least makes the code a bit simpler, and the byte order becomes more apparent.

Reply:
I can't seem to leave a comment on those lines, but lines 60-61 include the doc comment

> The size of an `Ipv4Addr` struct may vary depending on the target operating system

It seems like this can be removed, as it's no longer relevant and just plain incorrect given the new implementation.

A similar doc comment exists for `Ipv6Addr`

Reply:
> It seems like this can be removed, as it's no longer relevant and just plain incorrect given the new implementation.

I agree! I have commited this change and it will be in this PR shortly.